text/3537-msrv-resolver.md

```diff
@@ -0,0 +1,570 @@
+[prior-art]: #prior-art
+
+- Python: instead of tying packages to a particular tooling version, the community instead focuses on their equivalent of the [`rustversion` crate](https://crates.io/crates/rustversion) combined with tool-version-conditional dependencies that allow polyfills.
+  - We have [cfg_accessible](https://github.com/rust-lang/rust/issues/64797) as a first step though it has been stalled
+  - These don't have to be mutually exclusive solutions as conditional compilation offers flexibility at the cost of maintenance.  Different maintainers might make different decisions in how much they leverage each
+  - One big difference is Python continues to support previous releases which sets a standard within the community for "MSRV" policies.
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+The config field is fairly rought
+  - The location (within `build`) needs more consideration
+  - The name isn't very clear
+  - The values are awkward
+
+# Future possibilities
+[future-possibilities]: #future-possibilities
+
+## Improve the experience with lack of `rust-version`
+
+The user experience for this is based on the extent and quality of the data.
+Ensuring we have `package.rust-version` populated more often (while maintaining
+quality of that data) is an important problem but does not have to be solved to
+get value out of this RFC and can be handled separately.
+
+~~We could encourage people to set their MSRV by having `cargo new` default `package.rust-version`.~~
+However, if people aren't committed to verifying it,
+it is likely to go stale and will claim an MSRV much older than what is used in practice.
+If we had the hard-error resolver mode and
+[clippy warning people when using API items stabilized after their MSRV](https://github.com/rust-lang/rust-clippy/issues/6324),
+this will at least annoy people into either being somewhat compatible or removing the field.
+
+~~When missing, `cargo publish` could inject `package.rust-version` using the version of rustc used during publish.~~
+However, this will err on the side of a higher MSRV than necessary and the only way to
+workaround it is to set `CARGO_BUILD_RESOLVER_PRECEDENCE=maximum` which will then lose
+all other protections.
+
+~~When missing, `cargo publish` could inject based on the rustup toolchain file.~~
+However, this will err on the side of a higher MSRV than necessary as well.
+
+~~When missing, `cargo publish` could inject `package.rust-version` inferred from
+`package.edition` and/or other `Cargo.toml` fields.~~
+However, this will err on the side of too low of an MSRV.
+While this might help with in this situation,
+it would lock us in to inaccurate information which might limit what analysis we could do in the future.
+
+Alternatively, `cargo publish` / the registry could add new fields to the Index
+to represent an inferred MSRV, the published version, etc
+so it can inform our decisions without losing the intent of the publisher.
+
```

Comment:
```suggestion

We could help people keep their MSRV up to date, by letting them specify a policy (e.g. `rust-version-policy = "stable - 2"` or `rust-version-policy = "stable"`); then, every time the user runs `cargo update`, we could automatically update their `rust-version` field as well.

```

Reply:
A bit unsure where we'd source `stable` from and if `cargo update` is the right location but I'm good with the idea.

Reply:
Side note: I've wondered if a `stable % 2` might be an interesting approach to MSRV.  It reduces MSRV update churn and keeps things consistent for when a particular MSRV does work.  I've not experimented with it as I can't experiment with it with RenovateBot afaik

Reply:
@epage My premise was a ratchet: if a version of cargo that knows it's a stable version (with a stable rustc) looks at the file, it can say "well, stable is *at least* my version, and that would move the MSRV forward, so let me update the MSRV accordingly". That way, after someone does a `rustup update`, their next `cargo update` (or some other operation if you prefer) will bump MSRV. That helps people keep MSRV up to date with their documented policy.

Reply:
So to check my understanding, `stable` is not referring to the currently released `stable` according to `rustup` but is referring to `rustc --version` if no pre-release fields are present (and no MSRV update would happen on beta, nightly).  Is that right?